sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -189,13 +205,45 @@ final class DataFrameWriter private[sql](df: DataFrame) {
         ifNotExists = false)).toRdd
   }
 
-  private def normalizedParCols: Option[Seq[String]] = partitioningColumns.map { parCols =>
-    parCols.map { col =>
-      df.logicalPlan.output
-        .map(_.name)
-        .find(df.sqlContext.analyzer.resolver(_, col))
-        .getOrElse(throw new AnalysisException(s"Partition column $col not found in existing " +
-          s"columns (${df.logicalPlan.output.map(_.name).mkString(", ")})"))
+  private def normalizedParCols: Option[Seq[String]] = partitioningColumns.map { cols =>
+    cols.map(normalize(_, "Partition"))
+  }
+
+  private def normalizedBucketCols: Option[Seq[String]] = bucketingColumns.map { cols =>
+    cols.map(normalize(_, "Bucketing"))
+  }
+
+  private def normalizedSortCols: Option[Seq[String]] = sortingColumns.map { cols =>
+    cols.map(normalize(_, "Sorting"))
+  }
+
+  private def getBucketSpec: Option[BucketSpec] = {
+    if (numBuckets.isEmpty && sortingColumns.isDefined) {
+      throw new IllegalArgumentException(
+        "Specify bucketing information by bucketBy when use sortBy.")
+    }
+    if (numBuckets.isDefined && numBuckets.get <= 0) {
+      throw new IllegalArgumentException("numBuckets must be greater than 0.")
+    }
+
+    if (numBuckets.isDefined) {
+      Some(BucketSpec(numBuckets.get, normalizedBucketCols.get, normalizedSortCols))
+    } else {
+      None
+    }
```

Comment:
This method can be simplified to:

```scala
if (sortingColumns.isDefined) {
  require(numBuckets.isDefined, "sortBy must be used together with bucketBy")
}

for {
  n <- numBuckets
  cols <- normalizedBucketCols
} yield {
  require(n > 0, "Bucket number must be greater than 0.")
  BucketSpec(n, cols, normalizedSortCols)
}
```

(`require` throws `IllegalArgumentException` when the condition is not met.)